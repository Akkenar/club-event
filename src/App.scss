@import './style/mixins';

.App__Main {
  max-width: 900px;
  margin: 0 auto;
  min-height: calc(100vh - 10.5em);
}

.App__MainContainer {
  margin: 2em;
}
