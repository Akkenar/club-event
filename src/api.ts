import environment from './environment/environment';

export function getApi(): any {
  // Depending on whether the file is replaced by Webpack, the api URLs
  // will change based on the type of build. See `environment.ts` and
  // `environment.prod.ts` for more details.
  return environment.api;
}
