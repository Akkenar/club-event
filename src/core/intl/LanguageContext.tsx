import * as React from 'react';
import { IntlType } from './intl.type';

const LanguageContext = React.createContext<IntlType>({
  confirmation: '',
  information: '',
  messages: {},
});

export default LanguageContext;
