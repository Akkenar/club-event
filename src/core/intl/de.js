import messages from '../../../translations/de/messages.de.json';
import confirmation from '../../../translations/de/confirmation.de.md';
import information from '../../../translations/de/information.de.md';

const intl = {
  messages,
  confirmation,
  information,
};

export default intl;
