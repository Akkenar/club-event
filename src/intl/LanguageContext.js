import React from 'react';
import { DEFAULT_LANGUAGE } from './provideLanguageInRoute';

const LanguageContext = React.createContext({
  messages: {},
  language: DEFAULT_LANGUAGE,
});

export default LanguageContext;
