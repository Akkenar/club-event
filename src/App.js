import React, { Fragment, useContext, useEffect } from 'react';
import { BrowserRouter, Redirect, Route, Switch } from 'react-router-dom';
import Header from './header/Header';

import SignupPageAsync from './signup-page/SignupPageAsync';
import HomePageAsync from './home-page/HomePageAsync';
import { withIntlManager } from './intl/provideIntlManager';
import LanguageContext from './intl/LanguageContext';
import ConfirmationPageAsync from './confirmation-page/ConfirmationPageAsync';
import FooterAsync from './footer/FooterAsync';

import './App.critical.scss';
import './index.critical.scss';
import './index.scss';

const App = () => {
  const { language } = useContext(LanguageContext);

  return (
    <Fragment>
      <div id="top" tabIndex="-1" className="visually-hidden">
        Top
      </div>
      <BrowserRouter>
        <div className="App">
          <header className="App__Header">
            <Header />
          </header>
          <main className="App__Main">
            <div className="App__MainContainer">
              <Switch>
                <Redirect exact={true} from="/" to={`home/${language}`} />
                <Route path="/home/:locale" component={HomePageAsync} />
                <Route path="/register/:locale" component={SignupPageAsync} />
                <Route
                  path="/confirmation/:locale"
                  component={ConfirmationPageAsync}
                />
              </Switch>
            </div>
          </main>
          <footer className="App__Footer">
            <FooterAsync />
          </footer>
        </div>
      </BrowserRouter>
    </Fragment>
  );
};

export default withIntlManager(App);
