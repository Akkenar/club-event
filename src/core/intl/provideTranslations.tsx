import * as React from 'react';
import {
  BrowserRouter,
  match as matchType,
  Redirect,
  Route,
  Switch,
} from 'react-router-dom';
import { getDefaultLanguage } from './getDefaultLanguage';
import LanguageContext from './LanguageContext';
import { useTranslations } from './useTranslations';

export const BASE_URL_WITH_LANG = '/:language([a-z]{2})';

interface ProvideMessagesProps {
  match: matchType<{ language: string }>;
}

export default function provideTranslations(
  Component: React.ComponentType<any>,
) {
  /**
   * The provided, depends on the routing and the language.
   * @param match
   * @constructor
   */
  const TranslationsProvider = ({ match }: ProvideMessagesProps) => {
    const targetLanguage = match.params.language;
    const value = useTranslations(targetLanguage);

    return (
      <LanguageContext.Provider value={value}>
        <Component />
      </LanguageContext.Provider>
    );
  };

  // Wrap the translation provider in a router.
  return () => (
    <BrowserRouter>
      <Switch>
        <Route
          path={`${BASE_URL_WITH_LANG}/:page`}
          component={TranslationsProvider}
        />
        <Redirect to={`/${getDefaultLanguage()}/home`} />
      </Switch>
    </BrowserRouter>
  );
}
