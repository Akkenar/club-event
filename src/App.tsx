import React from 'react';
import { useContext } from 'react';
import AppRouting from './AppRouting';
import FooterAsync from './core/footer/FooterAsync';
import HeaderAsync from './core/header/HeaderAsync';
import LanguageContext from './core/intl/LanguageContext';
import provideTranslations from './core/intl/provideTranslations';

import './App.scss';
import './style/font.scss';
import './style/main.scss';
import './style/semantic.scss';

export const AppContainer = () => {
  const { language } = useContext(LanguageContext);
  return (
    <>
      <div id="top" tabIndex={-1} className="visually-hidden">
        Top
      </div>
      <div className="App">
        <header className="App__Header no-print">
          <HeaderAsync />
        </header>
        <main className="App__Main">
          <div className="App__MainContainer">
            {language ? <AppRouting /> : null}
          </div>
        </main>
        <footer className="App__Footer no-print">
          <FooterAsync />
        </footer>
      </div>
    </>
  );
};

const App = provideTranslations(AppContainer);

export default App;
