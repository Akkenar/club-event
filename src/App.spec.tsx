import * as React from 'react';
import { Fragment } from 'react';
import { BrowserRouter, NavLink, Route } from 'react-router-dom';
import { act, render, fireEvent } from '@testing-library/react';
import App, { AppContainer } from './App';
import { Results } from './core/api.type';
import LanguageContext from './core/intl/LanguageContext';
import { BASE_URL_WITH_LANG } from './core/intl/provideTranslations';
import { mockInterestObservable } from './test-utils/intersectObservable-utils.lib';
import { mockResponse } from './test-utils/test-utils.lib';

// It's easier to mock the languages so we don't have to change the tests
// if we change the wording.
jest.mock('./core/intl/importLanguage', () => ({
  importLanguage: () => Promise.resolve({ default: { messages: {} } }),
}));

// A fake menu to access the pages in the route.
const Menu = () => {
  return (
    <Fragment>
      <NavLink data-testid="goto-home" to="/en/home">
        Home
      </NavLink>
      <NavLink data-testid="goto-register" to="/en/register">
        Register
      </NavLink>
      <NavLink data-testid="goto-login" to="/en/login">
        Login
      </NavLink>
      <NavLink data-testid="goto-confirmation" to="/en/confirmation">
        Confirmation
      </NavLink>
      <NavLink data-testid="goto-registrations" to="/en/registrations">
        Registrations
      </NavLink>
    </Fragment>
  );
};

const mockLanguageContext = {
  confirmation: '',
  information: '',
  schedule: '<table></table>',
  language: 'en',
  messages: {},
};

const AppWithMockLanguageAndRoute = () => (
  <LanguageContext.Provider value={mockLanguageContext}>
    <BrowserRouter>
      <Fragment>
        <Route path={BASE_URL_WITH_LANG} component={AppContainer} />
        <Menu />
      </Fragment>
    </BrowserRouter>
  </LanguageContext.Provider>
);

describe('App', () => {
  let showAsyncElements: () => void;
  beforeEach(() => {
    showAsyncElements = mockInterestObservable();
  });

  it('should render the home page by default', async () => {
    const { findByTestId, baseElement } = render(<App />);
    act(() => {
      showAsyncElements();
    });

    // Wait for the mock Home Page to be rendered to have a good overview of the App component.
    await findByTestId('main-title');
    expect(baseElement).toMatchSnapshot();
  });

  it('should render the home page', async () => {
    const { findByTestId, getByTestId } = render(
      <AppWithMockLanguageAndRoute />,
    );
    fireEvent.click(getByTestId('goto-home'));

    act(() => {
      showAsyncElements();
    });

    const element = await findByTestId('main-title');
    expect(element.innerHTML).toContain('home.page.title');
  });

  it('should render the register page', async () => {
    const { getByTestId, findByTestId } = render(
      <AppWithMockLanguageAndRoute />,
    );
    fireEvent.click(getByTestId('goto-register'));

    act(() => {
      showAsyncElements();
    });

    const element = await findByTestId('main-title');
    expect(element.innerHTML).toEqual('register.page.title');
  });

  it('should render the confirmation page', async () => {
    const { getByTestId, findByTestId } = render(
      <AppWithMockLanguageAndRoute />,
    );
    fireEvent.click(getByTestId('goto-confirmation'));

    act(() => {
      showAsyncElements();
    });

    const element = await findByTestId('main-title');
    expect(element.innerHTML).toEqual('confirmation.page.title');
  });

  it('should render the login page', async () => {
    const { getByTestId, findByTestId } = render(
      <AppWithMockLanguageAndRoute />,
    );
    fireEvent.click(getByTestId('goto-login'));

    act(() => {
      showAsyncElements();
    });

    const element = await findByTestId('main-title');
    expect(element.innerHTML).toEqual('login.page.title');
  });

  it('should render the registrations page', async () => {
    // Because the page fetches the data upon arrival.
    mockResponse(Results.SUCCESS);

    const { findByTestId, getByTestId } = render(
      <AppWithMockLanguageAndRoute />,
    );
    fireEvent.click(getByTestId('goto-registrations'));

    act(() => {
      showAsyncElements();
    });

    const element = await findByTestId('main-title');

    expect(element.innerHTML).toEqual('registrations.page.title');
  });
});
