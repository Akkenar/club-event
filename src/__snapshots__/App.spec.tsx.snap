// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render the home page by default 1`] = `
<body>
  <div>
    <div
      class="visually-hidden"
      id="top"
      tabindex="-1"
    >
      Top
    </div>
    <div
      class="App"
    >
      <header
        class="App__Header no-print"
      >
        <div>
          <div
            class="ui inverted Header menu"
          >
            <a
              aria-current="page"
              class="active"
              color="white"
              href="/en/home"
            >
              <svg
                aria-label="logo.alt"
                class="Logo Logo--inverted  Header__logo"
                data-testid="lazyloaded-placeholder"
                height="45"
                viewBox="0 0 1024 1024"
                width="60"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d=""
                />
              </svg>
            </a>
            <a
              class="item"
              href="/en/register"
            >
              header.register
            </a>
            <div
              aria-expanded="false"
              class="ui dropdown link item right LanguageSwitcher"
              data-testid="language-switcher"
              role="listbox"
              tabindex="0"
            >
              <div
                aria-live="polite"
                class="text"
                role="alert"
              >
                current.language
              </div>
              <i
                aria-hidden="true"
                class="dropdown icon"
              />
              <div
                class="menu transition"
              >
                <a
                  class="item"
                  href="/fr/home"
                  role="option"
                  title="Français"
                >
                  FR
                </a>
                <a
                  class="item"
                  href="/de/home"
                  role="option"
                  title="Deutsch"
                >
                  DE
                </a>
                <a
                  class="item"
                  href="/it/home"
                  role="option"
                  title="Italianno"
                >
                  IT
                </a>
                <a
                  aria-current="page"
                  class="item active"
                  href="/en/home"
                  role="option"
                  title="English"
                >
                  EN
                </a>
              </div>
            </div>
          </div>
        </div>
      </header>
      <main
        class="App__Main"
      >
        <div
          class="App__MainContainer"
        >
          <div>
            <h1
              class="ui header"
              data-testid="main-title"
            >
              home.page.title
            </h1>
            <svg
              aria-label="logo.alt"
              class="Logo  Logo--big"
              data-testid="lazyloaded-placeholder"
              height="232"
              viewBox="0 0 1024 1024"
              width="320"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d=""
              />
            </svg>
            <div />
            <div
              data-testid="iframe-placeholder"
            >
              loading
            </div>
            <h2
              class="ui header"
            >
              home.page.register
            </h2>
            <div
              class="InformationSection__register-container"
            >
              <a
                class="ui black button full-width"
                data-testid="register-button"
                href="/en/register"
                role="button"
              >
                register.now
              </a>
            </div>
            <svg
              aria-label="Photo: Arnaud Conne"
              class="LazyImage "
              data-testid="lazyloaded-placeholder"
              height="212"
              viewBox="0 0 1024 1024"
              width="320"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d=""
              />
            </svg>
          </div>
        </div>
      </main>
      <footer
        class="App__Footer no-print"
      >
        <div>
          <div
            class="ui inverted Footer menu"
          >
            <div
              class="item"
            >
              © GSL 2019
            </div>
            <div
              class="item"
            >
              <a
                href="mailto:ad-sss-info@speleo-lausanne.ch"
              >
                Contact
              </a>
            </div>
          </div>
        </div>
      </footer>
    </div>
  </div>
</body>
`;
